src/Cafe/OS/libs/nsyshid/Dimensions.cpp

```diff
@@ -0,0 +1,1076 @@
+		DimensionsMini& figure = GetFigureByIndex(index);
+		if (figure.index == 255)
+		{
+			return false;
+		}
+		// When a figure is removed from the toypad, respond to the game with the pad they were removed from, their index,
+		// the direction (0x01 in byte 6 for removed) and their UID
+		std::array<uint8, 32> figureChangeResponse = {0x56, 0x0b, figure.pad, 0x00, figure.index, 0x01,
+													  figure.data[0], figure.data[1], figure.data[2],
+													  figure.data[4], figure.data[5], figure.data[6], figure.data[7]};
+		figure.Save();
+		figure.dimFile.reset();
+		figure.index = 255;
+		figure.pad = 255;
+		figure.id = 0;
+		memcpy(&figureChangeResponse[6], figure.data.data(), 7);
+		figureChangeResponse[13] = GenerateChecksum(figureChangeResponse, 13);
+		m_figureAddedRemovedResponses.push(figureChangeResponse);
+		return true;
+	}
+
+	bool DimensionsUSB::CreateFigure(fs::path pathName, uint32 id)
+	{
+		FileStream* dimFile(FileStream::createFile2(pathName));
+		if (!dimFile)
+		{
+			return false;
+		}
+		std::array<uint8, 0x2D * 0x04> fileData{};
+		RandomUID(fileData.data());
+		fileData[7] = id & 0xFF;
+
+		std::array<uint8, 7> uid = {fileData[0], fileData[1], fileData[2], fileData[4], fileData[5], fileData[6], fileData[7]};
+
+		// Only characters are created with their ID encrypted and stored in pages 36 and 37,
+		// as well as a password stored in page 43. Blank tags have their information populated
+		// by the game when it calls the write_block command.
+		if (id != 0)
+		{
+			const std::array<uint8, 16> figureKey = GenerateFigureKey(fileData);
+
+			std::array<uint8, 8> valueToEncrypt = {uint8(id & 0xFF), uint8((id >> 8) & 0xFF), uint8((id >> 16) & 0xFF), uint8((id >> 24) & 0xFF),
+												   uint8(id & 0xFF), uint8((id >> 8) & 0xFF), uint8((id >> 16) & 0xFF), uint8((id >> 24) & 0xFF)};
+
+			std::array<uint8, 8> encrypted = Encrypt(valueToEncrypt.data(), figureKey);
+
+			std::memcpy(&fileData[36 * 4], &encrypted[0], 4);
+			std::memcpy(&fileData[37 * 4], &encrypted[4], 4);
+
+			std::memcpy(&fileData[43 * 4], PWDGenerate(fileData).data(), 4);
+		}
+		else
+		{
+			// Page 38 is used as verification for blank tags
+			fileData[(38 * 4) + 1] = 1;
+		}
+
+		if (fileData.size() != dimFile->writeData(fileData.data(), fileData.size()))
+		{
+			delete dimFile;
+			return false;
+		}
+		delete dimFile;
+		return true;
+	}
+
+	void DimensionsUSB::GenerateRandomNumber(uint8* buf, uint8 sequence,
+											 std::array<uint8, 32>& replyBuf)
+	{
+		// Decrypt payload into an 8 byte array
+		std::array<uint8, 8> value = Decrypt(buf, std::nullopt);
+		// Seed is the first 4 bytes (little endian) of the decrypted payload
+		uint32 seed = uint32(value[3]) << 24 | uint32(value[2]) << 16 | uint32(value[1]) << 8 | uint32(value[0]);
+		// Confirmation is the second 4 bytes (big endian) of the decrypted payload
+		uint32 conf = uint32(value[4]) << 24 | uint32(value[5]) << 16 | uint32(value[6]) << 8 | uint32(value[7]);
```

Comment:
This comment applies to all instances where you manually change endianness of a value. Cemu has utility code for this. Also loading a little endian value does not require extra work. You can just cast the reference.

```suggestion
		uint32 seed = uint32be::from_bevalue((uint32&)value[0]);
		// Confirmation is the second 4 bytes (big endian) of the decrypted payload
		uint32 conf = (uint32&)&value[4]
```

Reply:
> Does this apply cross platform (for the little endian code), because I thought byte order was dependent on platform?

This is true, but Cemu already makes the assumption that it's running on a little endian platform so making it portable across platforms of different endianness is already a huge task and making the assumption in one more place doesn't hurt in my opinion.